### PR TITLE
Issue 21: Fixing Bookkeeper Cluster deletion using charts

### LINKED ISSUE #21: Bookkeeper Cluster does not get deleted via charts
### Description
The Bookkeeper Cluster does not get deleted while using helm charts.

### Importance
must-have

### Location
charts

### PR DESCRIPTION
**Pull request by SrishT**
Signed-off-by: SrishT <Srishti.Thakkar@dell.com>

### Change log description
The supported versions map was getting deleted before deletion of Bookkeeper Cluster because of which the cluster deletion was not successful.

### Purpose of the change
Fixes #21 

### What the code does
It mounts the supported versions map to a volume in the operator.

### How to verify it
Bookkeeper Cluster will be deleted successfully using charts. 
The following scenarios were tested :

- Not creating the versions map before installing the operator - In this case operator does not come up and following event is seen *(MountVolume.SetUp failed for volume "versions-volume" : configmap "bk-supported-versions-map" not found)*
- Version map present and installation done using charts 
- Version map present and installation done manually
- Tried installing BK Cluster with an invalid version (installation fails)
- Tried installing BK Cluster with a valid version (installation happens)
- Tried uninstalling the BK Cluster and BK Operator using charts
- Tried uninstalling the BK Cluster and BK Operator manually

When the BK Cluster was successfully installed following scenarios were tried
- Upgrading to a valid version and confirmed that the upgrade happens
- Upgrading to an invalid version and confirmed that the upgrade fails 
- Confirmed that all resources were removed at the time of uninstallation (both manually and using charts) as expected.

### CODE REVIEW

File: charts/bookkeeper-operator/templates/version_map.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        0.3.0:0.3.0,0.3.1,0.3.2
+        0.3.1:0.3.1,0.3.2
+        0.3.2:0.3.2
+        0.4.0:0.4.0
+        0.5.0:0.5.0,0.5.1,0.6.0,0.6.1,0.6.2,0.7.0
+        0.5.1:0.5.1,0.6.0,0.6.1,0.6.2,0.7.0
+        0.6.0:0.6.0,0.6.1,0.6.2,0.7.0
+        0.6.1:0.6.1,0.6.2,0.7.0
+        0.6.2:0.6.2,0.7.0
+        0.7.0:0.7.0
```
Comment (jkhalack) [inline charts/bookkeeper-operator/templates/version_map.yaml:18]: Could you please add the 0.7.1 version here?
It's not possible to deploy a 0.7.1 bk image at the moment.

Comment (pbelgundi) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: @jkhalack We won't be adding versions not yet released on open source.
You can add those to your config map yaml, if needed.
Charts and deploy files provided in p-operator repo are only for reference.

File: deploy/crds/crd.yaml [hunk 2]
```diff
@@ -19,7 +19,7 @@ spec:
   - name: Desired Version
     type: string
     description: The desired bookkeeper version
-    JSONPath: .status.TargetVersion
+    JSONPath: .spec.version
```
Comment (pbelgundi) [inline deploy/crds/crd.yaml:22]: Why are we changing this?

File: pkg/webhook/webhook.go [hunk 3]
```diff
@@ -69,16 +70,20 @@ func (pwh *bookkeeperWebhookHandler) mutateBookkeeperManifest(ctx context.Contex
 }
 
 func (pwh *bookkeeperWebhookHandler) mutateBookkeeperVersion(ctx context.Context, bk *bookkeeperv1alpha1.BookkeeperCluster) error {
-	configMap := &corev1.ConfigMap{}
-	err := pwh.client.Get(ctx, types.NamespacedName{Name: util.ConfigMapNameForBookieVersions(bk.Name), Namespace: bk.Namespace}, configMap)
+	// The key is the supported versions, the value is a list of versions that can be upgraded to.
+	supportedVersions := make(map[string]string)
+
+	file, err := os.Open("/tmp/config/keys")
```
Comment (pbelgundi) [inline pkg/webhook/webhook.go:76]: Code to read from file should be added to a separate method which can be invoked here. Lets keep the code as modular as possible.

**Comment by pbelgundi** [comment]
Please explain in detail all use-cases tested for this fix.
We need to make sure we don't break existing functionality when adding new features.

File: deploy/crds/crd.yaml (hunk 2 above)
Comment (SrishT) [reply deploy/crds/crd.yaml:22]: The desired version is showing as an empty field while doing manual installation. This will fix that.

**Comment by SrishT** [comment]
The various use-cases tested were the following

- Not creating the versions map before installing the operator - In this case operator does not come up and following event is seen *(MountVolume.SetUp failed for volume "versions-volume" : configmap "bk-supported-versions-map" not found)*
- Version map present and installation done using charts 
- Version map present and installation done manually
- Tried installing Bk cluster with an invalid version (installation fails)
- Tried installing Bk cluster with a valid version (installation happens)

After bookkeeper cluster was successfully installed following scenarios were tried
- Upgrading to a valid version and confirmed that the upgrade happens
- Upgrading to an invalid version and confirmed that the upgrade fails 
- Confirmed that all resources were removed at the time of uninstallation (both manually and using charts) as expected.

File: pkg/webhook/webhook.go (hunk 3 above)
Comment (SrishT) [reply pkg/webhook/webhook.go]: done

File: deploy/crds/crd.yaml (hunk 2 above)
Comment (pbelgundi) [reply deploy/crds/crd.yaml:22]: Lets not add this change here unless it is needed for fixing the version map issue.

File: pkg/webhook/webhook.go [hunk 4]
```diff
@@ -211,3 +208,19 @@ func (pwh *bookkeeperWebhookHandler) InjectScheme(s *runtime.Scheme) error {
 	pwh.scheme = s
 	return nil
 }
+
+func CreateVersionMap() (map[string]string, error) {
```
Comment (pbelgundi) [inline pkg/webhook/webhook.go:212]: please create a private method with starting letter in small case.

File: deploy/crds/crd.yaml (hunk 2 above)
Comment (SrishT) [reply deploy/crds/crd.yaml]: alright

File: pkg/webhook/webhook.go (hunk 4 above)
Comment (SrishT) [reply pkg/webhook/webhook.go]: done

File: charts/bookkeeper-operator/templates/version_map.yaml (hunk 1 above)
Comment (jkhalack) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: @pbelgundi Are you saying that this file is for example purposes only? 
Why is it placed into the operator chart then?

Comment (pbelgundi) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: @jkhalack Not sure why this should be so difficult to understand .... but charts and deploy files provided in the p-operator repository are for use by pravega opensource users and aligned to pravega opensource releases. If you choose to use different image versions of pravega you obviously would need to modify the .yaml to add those versions. That was the sole reason for bringing the version map outside the code and injecting as a config map.

Comment (jkhalack) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: @pbelgundi I don't understand the updated use case for the versions map.
Should it now have a fixed name `bk-supported-versions-map`, not related to the name of BK cluster (as it was assumed previously)?

Comment (SrishT) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: The updated use case is that the name of the map can be anything as long as the same file is mounted to the operator volume. Hence as opposed to earlier where we needed the name to be in a certain format, now we don't have any such requirement for the name of this configmap.

Comment (jkhalack) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: I meant the following block:
```
metadata:
  name: bk-supported-versions-map
```
Can it be anything?

Comment (jkhalack) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: And why is this conversation resolved if there is no answer to the original question?

Comment (jkhalack) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: @SrishT Are you saying that this `metaname.name` for configmap should be the same as line 44 of the operator chart? 
https://github.com/pravega/bookkeeper-operator/blob/master/charts/bookkeeper-operator/templates/operator.yaml#L44

Then we should not need the versions map that used to reside in `charts/bookkeeper/templates` folder. Is this correct?

Comment (SrishT) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: yes Julia that is correct. The version map will not be residing in `charts/bookkeeper/templates` any longer. Also we need to ensure that value in `metaname.name` should be same as the value mentioned in this line https://github.com/pravega/bookkeeper-operator/blob/master/charts/bookkeeper-operator/templates/operator.yaml#L44

Comment (jkhalack) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: Thank you, Srishti.
So the situation is that one cannot do any dev work on BK image without either
- downgrading the BK image version tag to `0.7.0`, or
- forking the current repository and making changes to bk-operator chart (note that https://github.com/pravega/bookkeeper-operator/blob/master/charts/bookkeeper-operator/templates/operator.yaml#L44 is not overridable at install time, so either that line has to change or the chart for `bk-supported-versions-map` configmap has to change).

Comment (SrishT) [reply charts/bookkeeper-operator/templates/version_map.yaml:18]: I will add `0.7.1` to this map in a PR and get it merged. Does that work @jkhalack

**Pull request merged**